R/Rules-methods.R

```diff
@@ -526,259 +521,136 @@ setMethod(
-          } else {
-            which.max(data@doseGrid[dosesOK])
-          }
+#' @describeIn nextBest find the next best dose based on the dual endpoint
+#'   model. The additional list element `probs` contains the target and
+#'   overdosing probabilities (across all doses in the dose grid) used in the
+#'   derivation of the next best dose.
+#'
+#' @aliases nextBest-NextBestDualEndpoint
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestDualEndpoint.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestDualEndpoint",
+    doselimit = "numeric",
+    samples = "Samples",
+    model = "DualEndpoint",
+    data = "Data"
+  ),
+  definition = function(nextBest, doselimit, samples, model, data, ...) {
+    doselimit <- ifelse(missing(doselimit) || length(doselimit) == 0, Inf, doselimit)
 
-        ret <- data@doseGrid[dosesOK][doseLevel]
+    # Biomarker samples at the dose grid points.
+    biom_samples <- samples@data$betaW
+
+    if (nextBest@target_relative) {
+      # If 'Emax' parameter available, target biomarker level will be relative to 'Emax',
+      # otherwise, it will be relative to the maximum biomarker level achieved
+      # in dose range for a given sample.
+      if ("Emax" %in% names(samples@data)) {
+        lwr <- nextBest@target[1] * samples@data$Emax
+        upr <- nextBest@target[2] * samples@data$Emax
+        prob_target <- colMeans(apply(biom_samples, 2L, function(s) (s >= lwr) & (s <= upr)))
       } else {
-        ## if none of the doses is OK:
-        doseLevel <- NA
-        ret <- NA
+        target_levels <- apply(biom_samples, 1L, function(x) {
+          rng <- range(x)
+          min(which(h_in_range(x, nextBest@target * diff(rng) + rng[1] + c(0, 1e-10), c(FALSE, TRUE))))
+        })
+        prob_target <- as.vector(table(factor(target_levels, levels = 1:data@nGrid)))
+        prob_target <- prob_target / nrow(biom_samples)
       }
+    } else {
+      prob_target <- colMeans(h_in_range(biom_samples, nextBest@target))
+    }
 
-      ## produce plot
-
-      ## first for the target probability
-      plot1 <- ggplot() +
-        geom_bar(
-          data =
-            data.frame(
-              x = data@doseGrid,
-              y = probTarget * 100
-            ),
-          aes(x = x, y = y),
-          stat = "identity",
-          position = "identity",
-          width = min(diff(data@doseGrid)) / 2,
-          colour = "darkgreen",
-          fill = "darkgreen"
-        ) +
-        xlab("Dose") +
-        ylab(paste("Target probability [%]")) +
-        ylim(c(0, 100))
+    # Now, compute probabilities to be in overdose tox interval, then flag
+    # eligible grid doses after accounting for maximum possible dose and discarding overdoses.
+    prob_samples <- sapply(data@doseGrid, prob, model = model, samples = samples)
+    prob_overdose <- colMeans(h_in_range(prob_samples, nextBest@overdose, c(FALSE, TRUE)))
```

Comment:
I think it would be a bit better readable if you use the argument name for the last logical argument

Reply:
Sure, the line will be a bit longer but I guess this is ok.